polynote-frontend/polynote/cell.js

```diff
@@ -855,4 +945,53 @@ export class TextCell extends Cell {
         // TODO: implement applyEdits for TextCell once rich text editor is figured out
         super.applyEdits(edits);
     }
+
+    getContentNodes() {
+        return Array.from(this.editorEl.childNodes)
+            // there are a bunch of text nodes with newlines we don't care about.
+            .filter(node => !(node.nodeType === Node.TEXT_NODE && node.textContent === '\n'))
+    }
+
+    // Note: lines in contenteditable are inherently weird, don't rely on this for anything aside from beginning and end
```

Comment:
@jeremyrsmith, after our conversation I was able to modify my existing code to get it to work reasonably well while still providing some semblance of position. I added this note to show that the lines, while (ostensibly) internally consistent, don't necessarily reflect anything you might experience as a user.